app/models/concerns/thredded/topic_common.rb

```diff
@@ -47,24 +47,45 @@ def unread(user)
           .merge(reads_class.where(reads[:id].eq(nil).or(reads[:read_at].lt(topics[:updated_at]))))
       end
 
+      private
+
+      # @param user [Thredded.user_class]
+      # @return [ByPostableLookup]
+      def read_states_by_topics_lookup(user, null_value)
+        read_states = reflect_on_association(:user_read_states).klass
+          .where(user_id: user.id, postable_id: current_scope.map(&:id))
+        Thredded::TopicCommon::Lookup.new(read_states, null_value)
```

Comment:
I'm coming in here with fresh eyes so maybe I can help with some outside perspective.

While the hash inject/reduce is certainly down closer to the metal (Ha! Ruby! Close to the metal!) it's a bit more difficult to grok. I agree that with some better naming that could be more readable but, for what it's worth, I tried to take that code snippet and expand it out with some different names and comments and I couldn't quite nail it. To be quite honest - I don't use map/reduce or collect/inject often. I'm not sure I'd be able to quite pull my thoughts together with this approach easily. I could deal if I had to, of course - but a lesser experienced rubyist *might* look at this and ["derp" themselves out the door](https://media4.giphy.com/media/1WQLKmc1Gfhny/200.gif).

I should probably admit, actually, my eyes are less than fresh after chasing my kid around for the evening. So - I'd probably vote for clarity and a well named, well factored class. The less I have to think, the better.

That being said. 

@timdiggins would you be averse to extracting the protected class into a name-spaced file and adding some tests? Considering the conversation here I'm coming mostly from the angle of a test providing both good documentation and exhibiting example behavior.